test/e2e/network/dual_stack_host_ips.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/network/common"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	netutils "k8s.io/utils/net"
+)
+
+var _ = common.SIGDescribe("Dual Stack [Feature:PodHostIPs]", func() {
+	f := framework.NewDefaultFramework("dualstack")
+
+	ginkgo.It("should create pod, add host ips is empty", func() {
+		e2eskipper.SkipIfFeatureGateEnabled(kubefeatures.PodHostIPs)
+
+		podName := "pod-dualstack-host-ips"
+
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   podName,
+				Labels: map[string]string{"test": "dualstack-host-ips"},
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:  "dualstack-host-ips",
+						Image: imageutils.GetE2EImage(imageutils.Agnhost),
+					},
+				},
+			},
+		}
+
+		ginkgo.By("submitting the pod to kubernetes")
+		podClient := f.PodClient()
+		p := podClient.CreateSync(pod)
+
+		gomega.Expect(p.Status.HostIP).ShouldNot(gomega.BeEquivalentTo(""))
+		gomega.Expect(p.Status.HostIPs).Should(gomega.BeNil())
```

Comment:
this is the test with the feature gate set, should not this be the one with HostIPs?

Reply:
This test is Disable FeatureGate, which should be empty.
The next test is Enable FeatureGate

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2681-pod-host-ip/README.md#expected-behavior

Reply:
this `e2eskipper.SkipUnlessFeatureGateEnabled(kubefeatures.PodHostIPs` doesn't work as you think :smile: 
The skipper doesn't have a way to check the code if the feature is enabled and depend on someone providing the right value in the e2e binary flags

For the e2e framework the convention is to use the tag to discriminate, `[Feature:PodHostIPs]` in the title means the feature is enabled, the absence means the feature status depends on the code base status (alpha disabled, beta: enabled, ga: enabled)

Reply:
I intend to test the enable, disable, and back.

but I don't found a tempSetCurrentKubeletConfig to reset the FeatureGate in this directory like the e2e_node directory 

Also, I hardcoded the turn on/off this Feature to test this enable and disable, then pass.

> For the e2e framework the convention is to use the tag to discriminate, [Feature:PodHostIPs] in the title means the feature is enabled, the absence means the feature status depends on the code base status (alpha disabled, beta: enabled, ga: enabled)

Does this mean that only the enabled state needs to be tested?

Reply:
e2e_node is different than e2e , it address different scenarios.
Think in e2e as a cluster that a user has already deployed, you can not have permutations on the inputs, the cluster is how the user has deployed, so you can only choose the tests that are going to run based on the test description.

Reply:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101566/pull-kubernetes-e2e-ubuntu-gce-network-policies/1505865988418048000

Look at the error message, it seems that it does not turn on Feature according to `[Feature:PodHostIPs]`.

Reply:
the feature has to be enabled manually in the cluster, for CI that means that the job configuration has to explicitly set the feature gate ...

see this job per example https://github.com/kubernetes/test-infra/blob/4d7b0d15c5b7551ecc9d0a23ddf1a5a9c696de1f/config/jobs/kubernetes/sig-gcp/sig-gcp-gce-config.yaml#L210-L232

Reply:
Does that mean, if I want this e2e pass, I have to add this in featureGate first?
Or should I remove this e2e first and add it in the bate stage?

Reply:
you can't do table tests here, so ` common.SIGDescribe("Dual Stack [Feature:PodHostIPs]` assumes it runs in a cluster with the feature gate set, for e2e just create the test assuming that the feature gate is enabled ... in addition, you may add one test in the e2e_node framework, since it seems it allows to use this pattern, but I'm not familiar with that framework